### PR TITLE
EN 4764 logger subsytem refactor

### PR DESCRIPTION
**Pull request by iulianpascalau**
added a new implementation for logger that follows ETH standard and for now it displays only on console the data

### CODE REVIEW

File: logger/commonConverters.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import (
+	"fmt"
+	"strings"
+	"time"
+)
+
+const msgFixedLength = 40
+
+func displayTime(time time.Time) string {
+	return fmt.Sprintf("%d-%02d-%02d %02d:%02d:%02d",
```
Comment (bogdan-rosianu) [inline logger/commonConverters.go]: I think this is hard to follow. I recommend this approach:
`return time.Format("2006-01-02 15:04:05")` 
it will return the same stuff and it's easier to change.
Also rename `time` parameter as it collides with a package name

File: logger/commonConverters_test.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+	"github.com/stretchr/testify/assert"
+)
+
+func TestFormatMessage_ShouldOutputFixedOrLargerStringThanMsgFixedLength(t *testing.T) {
+	t.Parallel()
+
+	testData := make(map[string]int)
+	testData[""] = 40
+	testData["small string"] = 40
+	largeString := "large stirng: " + strings.Repeat("A", msgFixedLength)
```
Comment (bogdan-rosianu) [inline logger/commonConverters_test.go]: large string*

File: logger/consoleFormatter.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+func displayArgs(levelColor string, args ...interface{}) string {
+	if len(args) == 0 {
+		return ""
+	}
+
+	argString := ""
+	for index := 1; index < len(args); index += 2 {
+		argString += fmt.Sprintf("\033[%s%s\033[0m=%v ", levelColor, args[index-1], args[index])
+	}
+	return argString
```
Comment (bogdan-rosianu) [inline logger/consoleFormatter.go]: add empty line

File: logger/consoleFormatter.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+// getLevelColor output the ANSI color code from the provided log level
+func getLevelColor(level LogLevel) string {
+	switch level {
+	case LogTrace:
+		return "0;37m"
```
Comment (bogdan-rosianu) [inline logger/consoleFormatter.go]: please define constants so we can know that `0;37m` is a kind of green for example

File: logger/examples_test.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+func generateHash() []byte {
+	buff := make([]byte, 32)
+	_, _ = rand.Reader.Read(buff)
+	return buff
+}
+
+//
+//hash := generateHash()
```
Comment (bogdan-rosianu) [inline logger/examples_test.go]: delete the commented code

File: logger/interface.go [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+	Trace(message string, args ...interface{})
+	Debug(message string, args ...interface{})
+	Info(message string, args ...interface{})
+	Warn(message string, args ...interface{})
+	Error(message string, args ...interface{})
+	LogIfError(err error, args ...interface{})
+	SetLevel(logLevel LogLevel)
+}
+
+// Formatter describes what an log formatter should be able to do
```
Comment (bogdan-rosianu) [inline logger/interface.go]: a log *

File: logger/logOutputSubject.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+func (los *logOutputSubject) AddObserver(w io.Writer, format Formatter) error {
+	if w == nil {
+		return ErrNilWriter
+	}
+	if format == nil {
+		return ErrNilFormatter
+	}
+
+	los.mutObservers.Lock()
+	defer los.mutObservers.Unlock()
```
Comment (bogdan-rosianu) [inline logger/logOutputSubject.go]: you can remove `defer` and move this after the protected lines

File: logger/logSubsystem.go [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+const ellipsisCharacter = "\u2026"
+
+var logMut = &sync.RWMutex{}
+var loggers map[string]*logger
+var defaultLogOut LogOutputHandler
+var defaultLogLevel = LogInfo
+
+func init() {
+	logMut.Lock()
+	defer logMut.Unlock()
```
Comment (bogdan-rosianu) [inline logger/logSubsystem.go]: remove `defer` and move this after the protected lines

File: logger/logger.go [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+	l.mutLevel.RLock()
+	shouldOutput := l.logLevel > compareLogLevel
+	l.mutLevel.RUnlock()
+
+	return shouldOutput
+}
+
+// Trace outputs a tracing log message with optional provided arguments
+func (l *logger) Trace(message string, args ...interface{}) {
+	if l.shouldOutput(LogTrace) {
```
Comment (bogdan-rosianu) [inline logger/logger.go]: this function and the next 3 are exactly the same beside of the log level. Maybe you can add a function a call that function from all these 4:
```GOLANG
func (l *logger) outputMessageFromLogLevel(level LogLevel, message string, args ...interface{}) {
   if l.shouldOutput(level) {
		return
   }

   logLine := newLogLine(message, level, args...)
   l.logOutput.Output(logLine)
}

File: logger/logSubsystem.go [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+func setLogLevelToVar(dest *LogLevel, pattern string, logLevel LogLevel) {
+	if pattern == "*" {
+		*dest = logLevel
+	}
+}
+
+// ParseLogLevelAndMatchingString can parse a string in the form "LOG_LEVEL|MATCHING_STRING" into its
+// corresponding log level and matching string. Errors if something goes wrong.
+// Example: "DEBUG|process" "INFO|*" "ERROR|process/sync" and so on
```
Comment (raduchis) [inline logger/logSubsystem.go]: Don't we want to be able to set multiple packages loglevels? --log_level=”state:info,mempool:error,*:error” 

Found the idea here, don't know how actual this is: https://blog.cosmos.network/one-of-the-exciting-new-features-in-0-10-0-release-is-smart-log-level-flag-e2506b4ab756

File: logger/interface.go [hunk 11]
```diff
@@ -0,0 +1,26 @@
+package logger
+
+import "io"
+
+// Logger defines the behavior of a data logger component
+type Logger interface {
+	Trace(message string, args ...interface{})
```
Comment (raduchis) [inline logger/interface.go]: Do we need a loglevel None? So that nothing is logged?

File: logger/logOutputSubject.go [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+
+	for i := 0; i < len(los.writers); i++ {
+		format := los.formatters[i]
+		buff := format.Output(line)
+		_, _ = los.writers[i].Write(buff)
+	}
+}
+
+// AddObserver adds a writer + formatter (called here observer) to the containing observer-like lists
+func (los *logOutputSubject) AddObserver(w io.Writer, format Formatter) error {
```
Comment (raduchis) [inline logger/logOutputSubject.go]: Could we need a "RemoveObserver" sometime? Also the AddObserver will not be used at runtime, no Observers could be added dynamically, when the app is running. Also no level could be set dynamically, while running.

File: logger/commonConverters.go (hunk 1 above)
Comment (iulianpascalau) [reply logger/commonConverters.go]: Done

File: logger/commonConverters_test.go (hunk 2 above)
Comment (iulianpascalau) [reply logger/commonConverters_test.go]: :D done

File: logger/consoleFormatter.go (hunk 3 above)
Comment (iulianpascalau) [reply logger/consoleFormatter.go]: done

File: logger/consoleFormatter.go (hunk 4 above)
Comment (iulianpascalau) [reply logger/consoleFormatter.go]: Done

File: logger/examples_test.go (hunk 5 above)
Comment (iulianpascalau) [reply logger/examples_test.go]: Done

File: logger/interface.go (hunk 6 above)
Comment (iulianpascalau) [reply logger/interface.go]: right

File: logger/logOutputSubject.go (hunk 7 above)
Comment (iulianpascalau) [reply logger/logOutputSubject.go]: Done, also on line 28

File: logger/logSubsystem.go (hunk 8 above)
Comment (iulianpascalau) [reply logger/logSubsystem.go]: done

File: logger/logger.go (hunk 9 above)
Comment (iulianpascalau) [reply logger/logger.go]: Done

File: logger/logSubsystem.go [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+	logMut.Lock()
+	defer logMut.Unlock()
+
+	loggers = make(map[string]*logger)
+	defaultLogOut = &logOutputSubject{}
+	_ = defaultLogOut.AddObserver(os.Stdout, ConsoleFormatter{})
+}
+
+// Get returns a log based on the name provided, generating a new log if there is no log with provided name
+func Get(name string) *logger {
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply logger/logSubsystem.go]: Right, sounds better. GetOrCreate shall be

File: logger/logSubsystem.go (hunk 10 above)
Comment (iulianpascalau) [reply logger/logSubsystem.go]: Right, good suggestion. Will modify

Comment (iulianpascalau) [reply logger/logSubsystem.go]: Done, changed to the above mentioned pattern.

File: logger/interface.go (hunk 11 above)
Comment (iulianpascalau) [reply logger/interface.go]: Done

File: logger/logOutputSubject.go (hunk 12 above)
Comment (iulianpascalau) [reply logger/logOutputSubject.go]: Right, added RemoveObserver function
